From ad2729965855da1ecf5360bd0da77bbadf6e2d41 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jakub=20Van=C4=9Bk?= <linuxtardis@gmail.com>
Date: Tue, 24 Feb 2026 22:04:36 +0100
Subject: [PATCH 1/3] net: phy: reverse MDIO bus scanning direction

Address 0 is incorrectly treated by some PHYs as a broadcast address.
Give these PHYs a chance to stop listening on that address by
probing addresses 1-31 before probing address 0.
---
 drivers/net/phy/mdio_bus.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

--- a/drivers/net/phy/mdio_bus.c
+++ b/drivers/net/phy/mdio_bus.c
@@ -614,7 +614,10 @@ static int mdiobus_scan_bus_c22(struct m
 {
 	int i;
 
-	for (i = 0; i < PHY_MAX_ADDR; i++) {
+	/* Probe address 0 last. Driver probe callbacks shall reconfigure
+	 * other PHYs that assume that address 0 is a broadcast address
+	 * before we get to address 0. */
+	for (i = PHY_MAX_ADDR - 1; i >= 0; i--) {
 		if ((bus->phy_mask & BIT(i)) == 0) {
 			struct phy_device *phydev;
 
@@ -630,7 +633,10 @@ static int mdiobus_scan_bus_c45(struct m
 {
 	int i;
 
-	for (i = 0; i < PHY_MAX_ADDR; i++) {
+	/* Probe address 0 last. Driver probe callbacks shall reconfigure
+	 * other PHYs that assume that address 0 is a broadcast address
+	 * before we get to address 0. */
+	for (i = PHY_MAX_ADDR - 1; i >= 0; i--) {
 		if ((bus->phy_mask & BIT(i)) == 0) {
 			struct phy_device *phydev;
 
